src/V3Timing.cpp

```diff
@@ -0,0 +1,1190 @@
+    // VISITORS
+    virtual void visit(AstClass* nodep) override {
+        VL_RESTORER(m_resetStmtsp);
+        VL_RESTORER(m_classScopep);
+        m_classp = nodep;
+        iterateChildren(nodep);
+        if (m_resetStmtsp) {
+            auto* const cancelEventsFuncp
+                = new AstCFunc{nodep->fileline(), "_cancel_events", m_classScopep, ""};
+            cancelEventsFuncp->argTypes(EmitCBaseVisitor::symClassVar());
+            cancelEventsFuncp->addStmtsp(m_resetStmtsp);
+            m_classScopep->addActivep(cancelEventsFuncp);
+            nodep->repairCache();
+        }
+    }
+    virtual void visit(AstScope* nodep) override {
+        if (m_classp && !m_classScopep) m_classScopep = nodep;
+        iterateChildren(nodep);
+    }
+    virtual void visit(AstVarScope* nodep) override {
+        if (!m_classp) return;
+        if (nodep->dtypep()->basicp() && nodep->dtypep()->basicp()->isEventValue()) {
+            AstNode* bodysp
+                = new AstText{nodep->fileline(), "vlSymsp->__Vm_eventDispatcher.cancel("};
+            bodysp->addNext(new AstVarRef{nodep->fileline(), nodep, VAccess::READ});
+            bodysp->addNext(new AstText{nodep->fileline(), ");\n"});
+            m_resetStmtsp = AstNode::addNext(m_resetStmtsp, bodysp);
+        }
+    }
+
+    //--------------------
+    virtual void visit(AstNode* nodep) override { iterateChildrenConst(nodep); }
+    virtual void visit(AstNodeModule*) override {}  // Accelarate
+
+public:
+    // CONSTRUCTORS
+    explicit TimingClassEventVisitor(AstNetlist* nodep) { iterate(nodep); }
+    virtual ~TimingClassEventVisitor() override {}
+};
+
+//######################################################################
+// Remove event triggers if not needed
+
+class TimingCleanTriggersVisitor final : public VNVisitor {
+private:
+    // METHODS
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // VISITORS
+    virtual void visit(AstEventTrigger* nodep) override {
+        nodep->replaceWith(new AstAssign{nodep->fileline(), nodep->trigp()->unlinkFrBack(),
+                                         new AstConst{nodep->fileline(), AstConst::BitTrue()}});
+        VL_DO_DANGLING(nodep->deleteTree(), nodep);
+    }
+
+    //--------------------
+    virtual void visit(AstNode* nodep) override { iterateChildrenConst(nodep); }
+
+public:
+    // CONSTRUCTORS
+    explicit TimingCleanTriggersVisitor(AstNetlist* nodep) { iterate(nodep); }
+    virtual ~TimingCleanTriggersVisitor() override {}
+};
+
+//######################################################################
+// Timing class functions
+
+void V3Timing::timingAll(AstNetlist* nodep) {
+    UINFO(2, __FUNCTION__ << ": " << endl);
+    UINFO(2, "  Transform Intra Assign Delays...\n");
+    { TimingIntraAssignControlVisitor{nodep}; }
+    V3Global::dumpCheckGlobalTree("timing_intra", 0, v3Global.opt.dumpTreeLevel(__FILE__) >= 6);
+    UINFO(2, "  Apply Timescale To Delays...\n");
+    { TimingDelayTimescaleVisitor{nodep}; }
```

Comment:
I'm skeptical you need this many netlist visitors, can't some be merged?  Or can you make a vector of everything to visit later (without doing whole netlist)? Each visitor adds multiple minutes in large designs.